jenkinsfiles: use shared pipelines

suskin opened the pull request:
## Overview

We have several jobs with the same pipeline code, so we want to be able
to reuse code across them. This uses Jenkins dynamic library loading and
execution to share pipeline code. This should make it easier to add and maintain the jobs.

## Testing done

I've tested the loading and executing here: https://jenkinsci.upbound.io/job/crossplane/job/stack-minimal-gcp/job/branch-create/job/test-submodule-build/5/console

## NOTE

Note that this shared code lives in a non-crossplane org repo; we should move it into a shared crossplane repo, but I don't have permissions to create one, so I'm using a separate one for now until that's created.

suskin committed:
jenkinsfiles: use shared pipelines

We have several jobs with the same pipeline code, so we want to be able
to reuse code across them. This uses Jenkins dynamic library loading and
execution to share pipeline code.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>

jbw976 reviewed (commented):
This definitely cleans up the Jenkinsfiles and reduces a lot of duplicated efforts.  That's really good @suskin!

Something I've really liked about the build submodule in core crossplane and the cloud provider stacks is that most all of the logic for building/publishing/testing/etc. is captured within that reusable submodule and only a minimal amount of logic is put into the Jenkinsfiles, enabling easy integration with any CI system and not just Jenkins.

It looks like we are doing something similar here where the Jenkinsfiles contain config and glue but most of the work is actually contained within the crossplane-cli commands.  The main exception is that we are doing tagging and branching within the Jenkinsfiles here, while core crossplane does that in the build submodule.

I think this is a good start that accomplishes reusability and minimal logic in the CI specific layer. Thanks @suskin!

jbw976 commented:
We can create a new repo for you to move your https://github.com/suskin/stack-cicd effort to @suskin.  Do you think `crossplane/stack-cicd` is the best name?  Another idea would be `crossplane/stack-build` similar to https://github.com/upbound/build.  Thoughts?

suskin commented:
> Do you think crossplane/stack-cicd is the best name? Another idea would be crossplane/stack-build

Well, currently the repo is specific to Jenkins, so that's why I originally named it using `cicd` instead of `build`. As in it can't really be used outside of Jenkins. It's possible that some other build stuff that **isn't** specific to Jenkins could go in there later, but it might also just go into the crossplane cli, so that's why I felt `stack-build` wasn't a clear winner. I'm fine with either one, but I'd lean toward using `stack-cicd` or even `stack-jenkins` for any Jenkins-specific work, and using `stack-build` for more general build utilities as we make them.

muvaf commented:
@suskin @jbw976 I feel like we've been using `stack-` prefix for repos that are actual `Stack`s. I'd lean towards dropping that for the new repo. Maybe `crossplane/cicd` or `crossplane/build`?

@suskin Other than that, I don't really know much about how Jenkins is set up, so, I'll defer to others for reviewing those PRs. Though it'd be great that if other repo's readme had some links to actual jenkins pipelins & explaining what works how and how people can use that in their own stacks.

displague reviewed (approved):
At first glance this change looks like a move from declarative jenkinsfile to scripted.  But I see in the suskin-cicd repo that the declarative style is fully preserved.

I wonder about the pattern of wrapping the entire pipeline in a single function, rather than modularizing each common set of steps.  I expected to see something more like that, where the `pipeline {}` for each `Jenkinsfile` was still available in each Stack. This would have looked more like: https://jenkins.io/blog/2017/02/15/declarative-notifications/#moving-the-code-to-the-library

Given that this pipeline will be heavily reused and unchanges, I am ok with this single function call pattern for Stack builds.

---

I would also prefer `crossplane/build` as the repo name.

suskin committed:
shared jobs: update repo url

We've moved the shared logic into a `crossplane` org repo, so we want
the jobs to point at that now.

Signed-off-by: suskin <813571+suskin@users.noreply.github.com>
